# Deribatives

Calculate derivatives symbolically.

In [5]:
from sympy import symbols, diff

## Informal definition of derivatives

The derivative of a function describes how the output of a function changes when there is a small change in an input variable.

For example, the cost $J$ is the output and $w$ is the input variable.

Let's give a 'small change' a name *epsilon* or $\epsilon$. We use these Greek letters because it is traditional in mathematics to use *epsilon*($\epsilon$) or *delta*($\Delta$) to represent a small value. You can think of it as representing 0.001 or some other small value.

$$
\begin{equation}
\text{if } w \uparrow \epsilon \text{ causes }J(w) \uparrow \text{by }k \times \epsilon \text{ then} \\

\frac{\partial J(w)}{\partial w} = k \tag{1}
\end{equation}
$$

This is just syas if you change the input to the function $J(w)$ by a little bit and the output changes by $k$ times that little bit, then the derivative of $J(w)$ is equal to $k$.

Let's try this out. Let's look at the derivative of the function $J(w) = w^2$ at the point $w=3$ and $\epsilon = 0.001$

In [7]:
J = (3)**2
J_epsilon = (3 + 0.001)**2
k = (J_epsilon - J)/0.001  # difference divided by epsilon
print(f"J = {J}, J_epsilon = {J_epsilon}, dJ_dw ~= k = {k:0.6f} ")

J = 9, J_epsilon = 9.006001, dJ_dw ~= k = 6.001000 


We have increased the input value a little bit (0.001), causing the output to change from 9 to 9.006001, an increase of 6 times the input increase. Referencing (1) above, this says that $k=6$, so $\frac{\partial J(w)}{\partial w} \approx 6$. If you are familar with calculus, you know, written symbolically, $\frac{\partial J(w)}{\partial w} = 2w$. With $w=3$ this is 6. Our calculation above is not exactly 6 because correct $\epsilon$ would need to be infinitesimally small or really, really small. That is why we use the symbols $\approx$ or ~= rather than +.

In [8]:
# Let's see what happens if we make epsilon smaller
J = (3)**2
J_epsilon = (3 + 0.000000001)**2
k = (J_epsilon - J)/0.000000001
print(f"J = {J}, J_epsilon = {J_epsilon}, dJ_dw ~= k = {k} ")

J = 9, J_epsilon = 9.000000006, dJ_dw ~= k = 6.000000496442226 


## Finding symbolic derivatives

In backdrop it is useful to know the derivative of simple functions at any input value. Put another way, we would like to know the 'symbolic' derivative rather than the 'arthmetic' derivative. An example of a symbolic derivative is, $\frac{\partial J(w)}{\partial w} = 2w$, the derivative of $J(w) = w^2$ above. With the symbolic derivative you can find the value the derivative at any input value $w$.

It turns out this process has been automated with symbolic differentation programs. An example of this in Python is the **SymPy** library.

### $J = w^2$

Define the python variables and their symbolic names.

In [9]:
J, w = symbols('J, w')

Define and print the expression. Note SymPy produces a latex string which 

In [12]:
J = w**2
J

w**2

Use SymPy's `diff` to differentiate the expression for $J$ with respect to $w$. Note the result matches our earlier example.

In [14]:
dJ_dw = diff(J, w)
dJ_dw

2*w

Evaluate the derivative at a few points by 'substituting' numeric values for the symbolic values.

In [16]:
dJ_dw.subs([(w, 2)])

4

In [18]:
dJ_dw.subs([(w, 3)])

6

In [19]:
dJ_dw.subs([(w, -3)])

-6

### $J=2w$

In [20]:
w, J = symbols('w, J')

In [21]:
J = 2 * w
J

2*w

In [23]:
dJ_dw = diff(J, w)
dJ_dw

2

In [24]:
dJ_dw.subs([(w, -3)])

2

### $J = \frac{1}{w^2}$

In [27]:
J, w = symbols('J, w')
J = 1/(w**2)
J

w**(-2)

In [28]:
dJ_dw = diff(J, w)

In [29]:
dJ_dw

-2/w**3

In [30]:
dJ_dw.subs([(w, 3)])

-2/27